### PR TITLE
nimble/ll: Fix updating adv and scan response data while advertising

### PR DESCRIPTION
**Pull request by rymanluk**
This PR make sure advertising is stopped while data are updating.

### CODE REVIEW

File: nimble/controller/src/ble_ll_adv.c [hunk 1]
```diff
@@ -3098,6 +3121,42 @@ ble_ll_adv_reschedule_event(struct ble_ll_adv_sm *advsm)
...
+
+#if MYNEWT_VAL(BLE_LL_CFG_FEAT_LL_EXT_ADV)
+    if (advsm->aux_active) {
+        return;
+    }
+#endif
+
+    if (advsm->flags & BLE_LL_ADV_SM_FLAG_NEW_ADV_DATA){
+
+        if (advsm->updated_data) {
```
Comment (andrzej-kaczmarek) [inline nimble/controller/src/ble_ll_adv.c]: why this `if` is needed here? flag should be enough here and also it's not present below

File: nimble/controller/src/ble_ll_adv.c [hunk 2]
```diff
@@ -3098,6 +3121,42 @@ ble_ll_adv_reschedule_event(struct ble_ll_adv_sm *advsm)
...
+        advsm->flags &= ~BLE_LL_ADV_SM_FLAG_NEW_ADV_DATA;
+    } else if (advsm->flags & BLE_LL_ADV_SM_FLAG_NEW_SCAN_RSP_DATA) {
+        os_mbuf_free_chain(advsm->scan_rsp_data);
+        advsm->scan_rsp_data = advsm->updated_data;
+        advsm->updated_data = NULL;
+        advsm->flags &= ~BLE_LL_ADV_SM_FLAG_NEW_SCAN_RSP_DATA;
+    }
+
+#if MYNEWT_VAL(BLE_LL_CFG_FEAT_LL_EXT_ADV)
+        /* DID shall be updated when host provides new advertising data */
```
Comment (andrzej-kaczmarek) [inline nimble/controller/src/ble_ll_adv.c]: indentation

File: nimble/controller/src/ble_ll_adv.c [hunk 3]
```diff
@@ -3098,6 +3121,42 @@ ble_ll_adv_reschedule_event(struct ble_ll_adv_sm *advsm)
...
+        return;
+    }
+
+#if MYNEWT_VAL(BLE_LL_CFG_FEAT_LL_EXT_ADV)
+    if (advsm->aux_active) {
+        return;
+    }
+#endif
+
+    if (advsm->flags & BLE_LL_ADV_SM_FLAG_NEW_ADV_DATA){
```
Comment (andrzej-kaczmarek) [inline nimble/controller/src/ble_ll_adv.c]: whitespace

File: nimble/controller/src/ble_ll_adv.c [hunk 4]
```diff
@@ -2001,16 +2004,26 @@ ble_ll_adv_set_scan_rsp_data(uint8_t *cmd, uint8_t cmd_len, uint8_t instance,
     new_data = (operation == BLE_HCI_LE_SET_EXT_SCAN_RSP_DATA_OPER_COMPLETE) ||
                (operation == BLE_HCI_LE_SET_EXT_SCAN_RSP_DATA_OPER_FIRST);
 
-    ble_ll_adv_update_data_mbuf(&advsm->scan_rsp_data, new_data,
-                                BLE_SCAN_RSP_DATA_MAX_LEN, cmd + 1, datalen);
-    if (!advsm->scan_rsp_data) {
-        return BLE_ERR_MEM_CAPACITY;
-    }
+    if (advsm->adv_enabled) {
+        assert(advsm->updated_data == NULL);
```
Comment (andrzej-kaczmarek) [inline nimble/controller/src/ble_ll_adv.c]: this is not ok - host is allowed to either update adv data twice in a row or just update both adv data and scan rsp data before adv event is finished

File: nimble/controller/src/ble_ll_adv.c [hunk 5]
```diff
@@ -105,6 +105,7 @@ struct ble_ll_adv_sm
     uint8_t initiator_addr[BLE_DEV_ADDR_LEN];
     struct os_mbuf *adv_data;
     struct os_mbuf *scan_rsp_data;
+    struct os_mbuf *updated_data;
```
Comment (andrzej-kaczmarek) [inline nimble/controller/src/ble_ll_adv.c]: not 100% sure, but this probably should be reset somewhere if adv is stopped before data were updated

File: nimble/controller/src/ble_ll_adv.c (hunk 1 above)
Comment (rymanluk) [reply nimble/controller/src/ble_ll_adv.c]: it is not enough for operation `Unchanged data (just update the Advertising DID)(0x04)`

File: nimble/controller/src/ble_ll_adv.c (hunk 4 above)
Comment (rymanluk) [reply nimble/controller/src/ble_ll_adv.c]: hmm, forgot about legacy advertising.

File: nimble/controller/src/ble_ll_adv.c (hunk 5 above)
Comment (rymanluk) [reply nimble/controller/src/ble_ll_adv.c]: not really reset but updated. done

**Comment by rymanluk** [comment]
@andrzej-kaczmarek fixed

**Pull request merged**